### DISCUSSION TITLE
Issue with the fog simulation code which is "attributeerror: Can't get attribute '_map' on <module '__mp_main__' "

### DISCUSSION THREAD

**Discussion by AIsimrand** [MartinHahner/LiDAR_fog_sim#10] [Q&A] [answered]
Hello Author,

I am working on generating synthetic data using your Lidar fog simulator on Kitti velodyne data. but somehow I am getting this particular error **"attributeerror: Can't get attribute 'map_test' on <module '__mp_main__' "** which I am unable to resolve or understand.
If you could just let me know any changes I need to do in the particular code would be great help for me.
Looking forward for your reply.
Thank you in advance

**Comment by MartinHahner** [comment, marked as answer]
Are you using Windows? It might be related to [that](https://stackoverflow.com/a/68883612).

**Reply by AIsimrand** [reply]
I am using mac os and the issue still persist.
could you tell me for generating fog simulation what all changes I need to do in fog_simulator.py code file?
that can be helpful for me 
Thank you!

**Reply by MartinHahner** [reply]
Sorry, I can't help you here. 
This error is not caused by a "bug" in the code, the error is raised because of the specific environment, you are using.
The code has been tested successfully on multiple settings: https://github.com/MartinHahner/LiDAR_fog_sim#disclaimer

[This](https://github.com/MartinHahner/LiDAR_fog_sim/files/8793889/foggy_lidar.txt) is the environment I am using on my Linux machine.
Here is the output of **pip list** in my environment:

```bash
Package             Version
------------------- -------------------
addict              2.4.0
anyio               3.3.1
argon2-cffi         21.1.0
attrs               21.2.0
Babel               2.9.1
backcall            0.2.0
black               22.1.0
bleach              4.1.0
cachetools          5.0.0
ccimport            0.3.7
certifi             2020.12.5
cffi                1.14.6
charset-normalizer  2.0.12
click               8.0.4
cumm-cu114          0.2.8
cycler              0.10.0
Cython              0.29.28
debugpy             1.4.3
decorator           5.1.0
defusedxml          0.7.1
deprecation         2.1.0
easydict            1.9
entrypoints         0.3
fire                0.4.0
flake8              4.0.1
idna                3.3
imageio             2.16.1
iniconfig           1.1.1
ipykernel           6.4.1
ipython             7.27.0
ipython-genutils    0.2.0
ipywidgets          8.0.0a6
jedi                0.18.0
Jinja2              3.0.1
joblib              1.1.0
json5               0.9.6
jsonschema          4.0.0a6
jupyter-client      7.0.2
jupyter-core        4.8.0rc1
jupyter-packaging   0.10.4
jupyter-server      1.11.0
jupyterlab          3.1.12
jupyterlab-pygments 0.1.2
jupyterlab-server   2.8.1
jupyterlab-widgets  2.0.0a3
kiwisolver          1.3.1
lark                1.1.2
llvmlite            0.38.0
lyft-dataset-sdk    0.0.8
MarkupSafe          2.0.1
matplotlib          3.3.4
matplotlib-inline   0.1.3
mccabe              0.6.1
mistune             0.8.4
mypy-extensions     0.4.3
nbclassic           0.3.1
nbclient            0.5.4
nbconvert           6.1.1b0
nbformat            5.1.3
nest-asyncio        1.5.1
networkx            2.7.1
ninja               1.10.2.3
notebook            6.4.3
numba               0.55.1
numpy               1.20.2
numpy-quaternion    2021.4.5.14.42.35
olefile             0.46
open3d              0.13.0+f7367a4
opencv-python       4.5.5.62
packaging           21.3
pandas              1.2.4
pandocfilters       1.5.0
parso               0.8.2
pathspec            0.9.0
pccm                0.3.4
pexpect             4.8.0
pickleshare         0.7.5
Pillow              8.3.2
pip                 21.0.1
platformdirs        2.5.1
plotly              5.6.0
pluggy              1.0.0
plyfile             0.7.3
portalocker         2.4.0
prometheus-client   0.11.0
prompt-toolkit      3.0.20
ptyprocess          0.7.0
py                  1.11.0
pybind11            2.9.1
pycodestyle         2.8.0
pycparser           2.20
pyflakes            2.4.0
Pygments            2.10.0
PyMieScatt          1.8.1.1
PyOpenGL            3.1.5
pyparsing           2.4.7
PyQt5               5.12.3
PyQt5-sip           4.19.18
PyQtChart           5.12
pyqtgraph           0.12.1
PyQtWebEngine       5.12.1
pyquaternion        0.9.9
pyrsistent          0.18.0
pytest              7.0.1
python-dateutil     2.8.1
python-pcl          0.2
pytz                2021.1
PyWavelets          1.2.0
PyYAML              6.0
pyzmq               22.2.1
requests            2.26.0
requests-unixsocket 0.2.0
scikit-image        0.19.2
scikit-learn        1.0.2
scipy               1.5.3
Send2Trash          1.8.1b0
setuptools          52.0.0.post20210125
Shapely             1.8.1.post1
SharedArray         3.2.1
sip                 4.19.13
six                 1.15.0
sklearn             0.0
sniffio             1.2.0
spconv-cu114        2.1.21
tenacity            8.0.1
termcolor           1.1.0
terminado           0.12.1
testpath            0.5.0
threadpoolctl       3.1.0
tifffile            2022.2.9
tomli               2.0.1
tomlkit             0.7.2
torch               1.10.2
tornado             6.1
tqdm                4.60.0
traitlets           5.1.0
typing-extensions   4.1.1
urllib3             1.26.8
wcwidth             0.2.5
webencodings        0.5.1
websocket-client    1.2.1
wheel               0.36.2
widgetsnbextension  4.0.0a4

```

**Reply by AIsimrand** [reply]
yeah I had read somewhere that the multiprocessing libraries mainly cause this issue and to resolve it I had to use another library o.e "multiprocess" it is same as of multiprocessing.
but now I am getting different error which is "**ValueError: cannot reshape array of size 461536 into shape (5)**" . Previously I had the same error while using pointcloud_viewer.py file which was solved by giving correct path to dataset. So now that I am getting the same error in fog_simulation.py file too I changed the path to data in that code file but its still showing the same error. I just want to know if I have to change path anywhere else in fog_simulation code other that Lidar_folder variable which is line 18 in code?
As I am using KITTI velodyne I have given path as **LIDAR_FOLDERS = ['training/velodyne','testing/velodyne']** so is this correct format and where else do I need to change path?

**Reply by MartinHahner** [reply]
Try changing **n_features** to **4**. KITTI might not have the channel encoded in the point cloud and only contains 4 features (x,y,z,i).
https://github.com/MartinHahner/LiDAR_fog_sim/blob/61b721b2c0574fb0b49ab265a73f0f38e4d0901f/fog_simulation.py#L28

**Reply by AIsimrand** [reply]
Thank you Author, 
the code runs fine now.
just 1 more question why is the fog_simulation code saving multiple velodyne_CVL_beta folder consisting of similar bin files?

**Reply by MartinHahner** [reply]
I am sorry, I don't know what you are talking about. Please be more specific.

**Reply by AIsimrand** [reply]
Yeah , 
well I am trying to run the fog_simulation.py script so its saving multiple folders that consist of .bin files. 
folders are named like **velodyne_CVL_beta_0.005** , **velodyne_CVL_beta_0.010** and so on.
So i just want to know that what is **available_alpha** as it's value keeps increasing and as per that it saves augmented fog ?

**Reply by MartinHahner** [reply]
Alpha ranges from 0.005 to 0.2, the values are taken from the files located here: 
https://github.com/MartinHahner/LiDAR_fog_sim/tree/main/integral_lookup_tables/original

https://github.com/MartinHahner/LiDAR_fog_sim/blob/61b721b2c0574fb0b49ab265a73f0f38e4d0901f/fog_simulation.py#L37-L49

The resulting bin files are the augmented point clouds containing different levels of fog. 
The higher alpha/beta, the denser the fog.

Please make sure to read and understand the paper before asking more questions. Thank you.
